pkg/sdn/plugin/dns.go

```diff
@@ -0,0 +1,237 @@
+	// Protects dnsMap operations
+	lock sync.Mutex
+	// Holds dns name and its corresponding information
+	dnsMap map[string]dnsValue
+}
+
+func CheckDNSResolver() error {
+	if _, err := exec.LookPath(DIG); err != nil {
+		return fmt.Errorf("%s is not installed", DIG)
+	}
+	return nil
+}
+
+func (d *DNS) Size() int {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	return len(d.dnsMap)
+}
+
+func (d *DNS) Get(dns string) dnsValue {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	data := dnsValue{}
+	if res, ok := d.dnsMap[dns]; ok {
+		data.ips = make([]net.IP, len(res.ips))
+		copy(data.ips, res.ips)
+		data.ttl = res.ttl
+		data.nextQueryTime = res.nextQueryTime
+	}
+	return data
+}
+
+func (d *DNS) Add(dns string) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if len(d.dnsMap) == 0 {
+		d.dnsMap = map[string]dnsValue{}
+	}
+	d.dnsMap[dns] = dnsValue{}
+	d.updateTTL(dns)
+	return d.updateIPs(dns)
+}
+
+func (d *DNS) Update() (bool, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	changed := false
+	errs := []error{}
+	for dns, oldres := range d.dnsMap {
+		if err := d.updateIPs(dns); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		if !equal(oldres.ips, d.dnsMap[dns].ips) {
+			changed = true
+		}
+	}
+	return changed, kerrors.NewAggregate(errs)
+}
+
+func (d *DNS) SyncTTL() bool {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	updated := false
+	for dns := range d.dnsMap {
+		if d.updateTTL(dns) {
+			updated = true
+		}
+	}
+	return updated
+}
+
+func (d *DNS) setTTL(dns string, ttl time.Duration) bool {
+	res, ok := d.dnsMap[dns]
+	if !ok {
+		// Should not happen, all operations on dnsMap are synchronized by d.lock
+		glog.Warningf("DNS value not found in dnsMap for domain: %q", dns)
+		return false
+	}
+
+	if res.ttl.Seconds() == ttl.Seconds() {
+		return false
+	}
+	res.ttl = ttl
+	d.dnsMap[dns] = res
+	return true
+}
+
+func (d *DNS) updateTTL(dns string) bool {
+	// Cached/Non-authoritative DNS servers only holds the remaining TTL for it to refresh but not the actual absolute TTL for the domain.
+	// Authoritative DNS servers holds actual TTL for the domain. So we need to:
+	// (a) Get all authorirative name servers for the given domain
+	// (b) Query TTL for the domain against one of the name servers from (a)
+
+	out, err := exec.Command("sh", []string{"-c", fmt.Sprintf("echo $(%s | head -n 1 | sed 's/.$//')", digShellCmdGetAuthoritativeNS(dns))}...).Output()
+	if err != nil || len(out) == 0 {
+		if err != nil {
+			glog.Errorf("Failed to find authoritative name servers for domain: %q, err: %v, defaulting ttl=%s", dns, err, defaultTTL.String())
+		} else {
+			glog.Warningf("Could not find authoritative name servers for domain: %q, defaulting ttl=%s", dns, defaultTTL.String())
+		}
+		return d.setTTL(dns, defaultTTL)
+	}
+	nameServer := strings.Trim(string(out[:]), "\n")
+
+	out, err = exec.Command("sh", []string{"-c", fmt.Sprintf("echo $(%s | awk '{print $2}')", digShellCmdGetTTL(dns, nameServer))}...).Output()
```

Comment:
Good point. Right now, current approach is fetch TTL once from auth server, Update IPs for every TTL from non-auth server and another routine that updates TTL periodically from auth servers.

We could simply this to fetch and update TTL and IP from non-auth server and perform the above step again based on last obtained TTL. In this case, we don't need to contact auth server and no need of another routine to update TTL.